File: libaries/ts/src/specialized/multiplayer/multiplayer.ts
import { RawServer } from "../../raw";
import { NetworkObject } from "../types";
import { Vector3, Vector2 } from "./positions";
import { Player } from "./player";

class MultiplayerServer extends RawServer {
    public players: Player[] = [];

    constructor(url: string, auth: string) {
        super(url, auth);
    }

    public async createPlayer(path: string, position: Vector3 | Vector2): Promise<Player> {
        let player = new Player(this, path, position);
        await player.set();
        this.players.push(player);
        return player;
    }

    public async pushAll(): Promise<void> {
        for (let player of this.players) {
            await player.set();
        }
    }
}

export { MultiplayerServer };

File: libraries/cpp/src/cpp/multiplayer.cpp
#include "../include/multiplayer.h"
#include "../include/raw.h"

Player::Player(RawServer *server, std::string path, Vector *position) {
    this->server = server;
    this->path = path;
    this->position = position;

    this->set();
}

void Player::set() {
    this->server->set(this->path, std::to_string(this->position->x) + "," + std::to_string(this->position->y) + "," + std::to_string(this->position->z));
}

Vector::Vector(float x, float y, float z) {
    this->x = x;
    this->y = y;
    this->z = z;
}


MultiplayerServer::MultiplayerServer(std::string url, std::string auth) : RawServer(url, auth) {
    this->url = url;
    this->auth = auth;
}

